File: scripts/run_examples_smoke.py
#!/usr/bin/env python3

from __future__ import annotations

import subprocess
import sys
from pathlib import Path


def main() -> int:
    repo_root = Path(__file__).resolve().parents[1]
    examples_dir = repo_root / "examples"

    example_paths = sorted(examples_dir.glob("*.js"))
    if not example_paths:
        print("No examples found under examples/*.js", file=sys.stderr)
        return 1

    binary_path = repo_root / "bin" / "giavascript-smoke"
    binary_path.parent.mkdir(parents=True, exist_ok=True)

    build_result = subprocess.run(
        ["crystal", "build", "src/giavascript_cli.cr", "-o", str(binary_path)],
        cwd=repo_root,
        text=True,
        capture_output=True,
    )
    if build_result.returncode != 0:
        sys.stderr.write(build_result.stdout + build_result.stderr)
        return build_result.returncode

    failures = []
    for example_path in example_paths:
        relative_example = example_path.relative_to(repo_root)
        print(f"Running {relative_example}")
        result = subprocess.run(
            [str(binary_path), str(relative_example)],
            cwd=repo_root,
            text=True,
            capture_output=True,
        )

        if result.returncode != 0:
            failures.append(relative_example)
            sys.stderr.write(result.stdout + result.stderr)

    if failures:
        print("\nExample smoke tests failed:", file=sys.stderr)
        for failed in failures:
            print(f"- {failed}", file=sys.stderr)
        return 1

    print(f"All {len(example_paths)} example smoke tests passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/run_examples_smoke.js
var listing = process.run("crystal", [
  "eval",
  "Dir.mkdir_p(\"bin\"); Dir.glob(\"examples/*.js\").sort.each { |path| puts path }"
]);

if (listing.status != 0) {
  console.error(listing.stdout + listing.stderr);
  process.exit(listing.status);
}

var examples = listing.stdout.trim().split("\n");
if (examples.length == 1 && examples[0] == "") {
  console.error("No examples found under examples/*.js");
  process.exit(1);
}

var binary = "bin/giavascript-smoke";
var build = process.run("crystal", [
  "build",
  "src/giavascript_cli.cr",
  "-o",
  binary
]);

if (build.status != 0) {
  console.error(build.stdout + build.stderr);
  process.exit(build.status);
}

var failures = [];
var result;
for (var example of examples) {
  console.log("Running " + example);
  result = process.run(binary, [example]);

  if (result.status != 0) {
    failures.push(example);
    console.error(result.stdout + result.stderr);
  }
}

if (failures.length > 0) {
  console.error("\nExample smoke tests failed:");
  for (var failure of failures) {
    console.error("- " + failure);
  }
  process.exit(1);
}

console.log("All " + examples.length + " example smoke tests passed");
process.exit(0);
